vllm/distributed/device_communicators/shm_broadcast.py

```diff
@@ -0,0 +1,145 @@
+from torch.distributed import ProcessGroup
+
+
+class ShmRingBuffer:
+
+    def __init__(self, pg: ProcessGroup, max_chunk_bytes, max_chunks):
+        self.rank = dist.get_rank(pg)
+        self.world_size = dist.get_world_size(pg)
+        global_ranks = dist.get_process_group_ranks(pg)
+        self.is_writer = self.rank == 0
+        self.is_reader = not self.is_writer
+        self.current_idx = 0
+        self.max_chunk_bytes = max_chunk_bytes
+        self.max_chunks = max_chunks
+        total_bytes = (self.max_chunk_bytes +
+                       self.world_size) * self.max_chunks
+        self.data_offset = 0
+        self.metadata_offset = self.max_chunk_bytes * self.max_chunks
+
+        if self.is_writer:
+            self.shared_memory = shared_memory.SharedMemory(create=True,
+                                                            size=total_bytes)
+            # initialize the metadata section to 0
+            for i in range(self.metadata_offset, total_bytes):
+                self.shared_memory.buf[i] = 0
+            dist.broadcast_object_list([self.shared_memory.name],
+                                       src=global_ranks[0])
+        else:
+            recv = [None]
+            dist.broadcast_object_list(recv, src=global_ranks[0])
+            name = recv[0]
+            # fix to https://stackoverflow.com/q/62748654/9191338
+            # Python incorrectly tracks shared memory even if it is not
+            # created by the process. The following patch is a workaround.
+            with patch("multiprocessing.resource_tracker.register",
+                       lambda *args, **kwargs: None):
+                self.shared_memory = shared_memory.SharedMemory(name=name)
+
+    @property
+    def data(self):
+        start = self.data_offset + self.current_idx * self.max_chunk_bytes
+        end = start + self.max_chunk_bytes
+        return memoryview(self.shared_memory.buf[start:end])
+
+    @property
+    def metadata(self):
+        start = self.metadata_offset + self.current_idx * self.world_size
+        end = start + self.world_size
+        return memoryview(self.shared_memory.buf[start:end])
+
+    @contextmanager
+    def acquire_write(self):
+        assert self.is_writer, "Only writers can acquire write"
+        start_index = self.current_idx
+        while True:
+            with self.metadata as metadata_buffer:
+                read_count = sum(metadata_buffer[1:])
+                written_flag = metadata_buffer[0]
+                if written_flag and read_count != self.world_size - 1:
+                    # this block is written and not read by all readers
+                    # try to write to the next block
+                    self.current_idx = (self.current_idx + 1) % self.max_chunks
+                    if self.current_idx == start_index:
+                        # no empty block found
+                        # wait for a while (0.1 us)
+                        time.sleep(1e-7)
```

Comment:
We may consider maintaining an accumulator and show a warning if this loop is stuck for an unreasonable long time, so that we could avoid silently hanging.

Reply:
good point. what's a reasonable timeout? for pytorch, it is:

> Default value is 10 minutes for NCCL and 30 minutes for other backends

Reply:
can we just print warning, instead of raising a timeout warning?

Reply:
Yeah print a warning would be sufficient I think. Maybe we can use 10 mins or even shorter, given that this communication itself doesn't have much overheads.

Reply:
fixed in [681919a](https://github.com/vllm-project/vllm/pull/5399/commits/681919a3a6a44904c5f3cac8edf23e832ee7a162)